unpublish.php

```diff
@@ -195,9 +196,36 @@ public function action_save_unpublish_timestamp( $post_id ) {
 		$unpublish_date = vsprintf( '%04d-%02d-%02d %02d:%02d:00', $date_parts );
 		$valid_date     = wp_checkdate( $date_parts['mm'], $date_parts['jj'], $date_parts['aa'], $unpublish_date );
 
-		if ( $valid_date ) {
-			update_post_meta( $post_id, self::$post_meta_key, strtotime( $unpublish_date ) );
+		if ( ! $valid_date ) {
+			return;
 		}
+
+		$timestamp = strtotime( get_gmt_from_date( $unpublish_date ) );
+
+		update_post_meta( $post_id, self::$post_meta_key, $timestamp );
+		$this->schedule_unpublish( $post_id, $timestamp );
+	}
+
+	/**
+	 * Unpublish post
+	 *
+	 * Invoked by cron 'unpublish_post_cron' event.
+	 *
+	 * @param int $post_id Post ID.
+	 */
+	public function unpublish_post( $post_id ) {
+		wp_trash_post( $post_id );
```

Comment:
I'd add a check to ensure the unpublish date has passed and reschedule if not (similar to the WP Core check for publishing scheduled posts. WP Cron can be a little unreliable :)

In psuedo code:
```php
if ( $unpublish_date > time() ) {
  // reschedule event
  return false;
}
// trash it
```